File: algorithms/python/heaters/heaters.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/heaters/
# Author: Miao Zhang
# Date:   2021-02-15

class Solution:
    def findRadius(self, houses: List[int], heaters: List[int]) -> int:
        res = 0
        n = len(heaters)
        j = 0
        heaters.sort()
        houses.sort()
        for h in houses:
            while j < n - 1 and abs(heaters[j + 1] - h) <= abs(heaters[j] - h):
                j += 1
            res = max(res, abs(heaters[j] - h))
        return res

File: algorithms/cpp/heaters/heaters.cpp
// Source: https://leetcode.com/problems/heaters/
// Author: Miao Zhang
// Date:   2021-02-15

class Solution {
public:
    int findRadius(vector<int>& houses, vector<int>& heaters) {
        int n = heaters.size();
        int j = 0;
        int res = 0;
        sort(houses.begin(), houses.end());
        sort(heaters.begin(), heaters.end());
        for (int h: houses) {
            auto pos = lower_bound(heaters.begin(), heaters.end(), h);
            int dist1 = (pos == heaters.end()) ? INT_MAX: *pos - h;
            int dist2 = (pos == heaters.begin()) ? INT_MAX : h - *(--pos);
            res = max(res, min(dist1, dist2));
        }
        return res;
    }
};
